### ISSUE TITLE
ExpressionRunner class hierarchy

### ISSUE THREAD

**Issue by aheejin** [WebAssembly/binaryen#2797]
#2702 exposes functionalities of [`ExpressionRunner`](https://github.com/WebAssembly/binaryen/blob/2704a9d3b4225f16e2637f38ebb200d588fffd26/src/wasm-interpreter.h#L151) to Binaryen and C API. I find most usages there is about precomputing certain expression, and we have [`PrecomputeExpressionRunner`](https://github.com/WebAssembly/binaryen/blob/2704a9d3b4225f16e2637f38ebb200d588fffd26/src/passes/Precompute.cpp#L46) for that. But after #2702 some functionalities of `PrecomputeExpressionRunner` are duplicated in `ExpressionRunner` to make them exposed to Binaryen and C API, and I'm not very sure what the difference between `ExpressionRunner` and `PrecomputeExpressionRunner` is at this point.

So I'm not sure the functionalities of `ExpressionRunner` class hierarchy is very well defined now. Can't Binaryen and C API make use of `PrecomputeExpressionRunner` instead, if the main purpose of that is precomputing?

I might be mistaken, so please advise me if so :)

cc @dcodeIO @kripken @tlively

**Comment by dcodeIO** [comment]
Yeah, that's a refactor I'm not especially proud of. Essentially, what happened there is that most of `PrecomputeExpressionRunner` has been refactored into the base `ExpressionRunner`, since an earlier attempt to move `PrecomputeExpressionRunner` to `wasm-interpreter.h` as `StandaloneExpressionRunner` (extending `ExpressionRunner`) turned out to be overly confusing in review. So the suggestion of moving all of it into the base class was made and implemented. There's also the situation now where the base `ExpressionRunner` has some fields that are never used by the derived `RuntimeExpressionRunner`, which I mentioned on the original PR as it seems unfortunate.

Perhaps we should pick up the idea again to add a class in between, what previously was `StandaloneExpressionRunner` (`ContextAwareExpressionRunner`?), that only `PrecomputeExpressionRunner` and `CExpressionRunner` inherit from, but not `RuntimeExpressionRunner`?

```
ExpressionRunner
├ WhateverExpressionRunner
│  ├ PrecomputeExpressionRunner
│  └ CExpressionRunner
├─ ConstantExpressionRunner
└─ RuntimeExpressionRunner
```

**Comment by aheejin** [comment]
> since an earlier attempt to move PrecomputeExpressionRunner to wasm-interpreter.h as StandaloneExpressionRunner (extending ExpressionRunner) turned out to be overly confusing in review.

I didn't review that PR, so could you elaborate on what parts were overly confusing in doing that?

**Comment by dcodeIO** [comment]
Sure, the pivotal discussion was https://github.com/WebAssembly/binaryen/pull/2702#discussion_r398032692, then leading to that change.

**Comment by aheejin** [comment]
Sorry, I read the comment and still at a loss.
- What does it mean to be the execution to be _linear_?
- Why is the comment thread a reason that Binaryen and C API cannot use `PrecomputeExpressionRunner` instead?

**Comment by dcodeIO** [comment]
> What does it mean to be the execution to be linear?

This has implications on the way we track intermediate values of locals (which is one of my additions that precompute didn't have previously). If execution was not be linear, we cannot simply use a map that we write to on `local.set`, and read on `local.get`. Somewhat analogous to the runner exclusively operating in execution order.

> Why is the comment thread a reason that we can't expose PrecomputeExpressionRunner instead so Binaryen and C API can use it?

See also https://github.com/WebAssembly/binaryen/pull/2804#issuecomment-619067162 on the other thread. Mostly boils down to the precompute pass having one special mechanism.

**Comment by kripken** [comment]
I'd support a refactoring here, I agree this should be cleaner. Hopefully we can find a nice solution!

Sorry I didn't push on this more in reviewing that PR. Part of the reason is I had "tunnel vision" in a large and long-running review of a PR - over time in review I focused less on the big picture and more on specific details. But in a big refactoring like that I should have re-read the entire thing more frequently.

**Comment by dcodeIO** [comment]
> Sorry I didn't push on this more in reviewing that PR. Part of the reason is I had "tunnel vision"

Can we please agree that this is all my fault for being terrible at C++ and too inexperienced with Binaryen's codebase, making my PR's hard to review in the first place? :) I promise this will improve a lot over time, when I become more confident! (I guess it also takes a little longer than usual since I'm external and mostly relying on somewhat indirect, so sometimes inefficient means of communication.)

**Comment by aheejin** [comment]
Addressed by #2804. Thanks!
